l2gersync/l2_ger_syncer.go

```diff
@@ -0,0 +1,173 @@
+			l2Client, l2GERManagerAddr,
+			l1InfoTreeSync, processor,
+			rh, bf, waitForNewBlocksPeriod,
+		)
+
+	case SovereignChain:
+		downloader, err = newDownloaderSovereign(
+			l2Client, l2GERManagerAddr,
+			l1InfoTreeSync,
+			rh, bf, waitForNewBlocksPeriod,
+		)
+
+	default:
+		return nil, fmt.Errorf("unknown sync mode %s provided", syncMode)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	compatibilityChecker := compatibility.NewCompatibilityCheck(
+		requireStorageContentCompatibility,
+		downloader.RuntimeData,
+		processor)
+	driver, err := sync.NewEVMDriver(rdL2, processor, downloader, reorgDetectorID,
+		downloadBufferSize, rh, compatibilityChecker)
+	if err != nil {
+		return nil, err
+	}
+
+	return &L2GERSync{
+		driver:    driver,
+		processor: processor,
+	}, nil
+}
+
+// resolveSyncMode determines the synchronization mode based on deployed provided l2 ger manager contract
+func resolveSyncMode(ctx context.Context, address common.Address, backend bind.ContractBackend) (SyncMode, error) {
+	// Try sovereign chain ger manager
+	sovereignGERManager, err :=
+		globalexitrootmanagerl2sovereignchain.NewGlobalexitrootmanagerl2sovereignchain(address, backend)
+	if err == nil {
+		updater, err := sovereignGERManager.GlobalExitRootUpdater(&bind.CallOpts{Context: ctx})
+		if err == nil {
+			log.Debugf("Detected: GlobalExitRootManagerL2SovereignChain, GlobalExitRootUpdater = %s", updater.Hex())
+			return SovereignChain, nil
+		}
+
+		if !strings.Contains(err.Error(), gethvm.ErrExecutionReverted.Error()) {
+			return "", fmt.Errorf("unexpected error when checking l2 ger manager (sovereign chain): %w", err)
+		}
+	}
+
+	// Try with legacy ger manager
+	legacyGERManager, err := polygonzkevmglobalexitrootv2.NewPolygonzkevmglobalexitrootv2(address, backend)
+	if err == nil {
+		bridgeAddr, err := legacyGERManager.BridgeAddress(&bind.CallOpts{Context: ctx})
+		if err == nil {
+			log.Debugf("Detected: PolygonZkEVMGlobalExitRootV2, BridgeAddress = %s", bridgeAddr)
+			return Legacy, nil
+		}
+	}
+
+	return "", fmt.Errorf("could not determine l2 ger sync mode based on the ger manager contract@%s", address.Hex())
+}
+
+// Start initiates the synchronization process.
+func (s *L2GERSync) Start(ctx context.Context) {
+	s.driver.Sync(ctx)
+}
+
+// GetFirstGERAfterL1InfoTreeIndex returns the first GER after a specified L1 info tree index
+func (s *L2GERSync) GetFirstGERAfterL1InfoTreeIndex(
+	ctx context.Context, atOrAfterL1InfoTreeIndex uint32,
+) (GlobalExitRootInfo, error) {
+	return s.processor.GetFirstGERAfterL1InfoTreeIndex(ctx, atOrAfterL1InfoTreeIndex)
+}
+
+// GetInjectedGERsForRange retrieves all injected global exit roots within a specified block range.
+// It returns a map where the keys are the global exit root hashes and the values are the
+// corresponding GlobalExitRootInfo containing the L1 info tree index, global exit root and block number.
+func (s *L2GERSync) GetInjectedGERsForRange(ctx context.Context,
```

Comment:
I think this is now not used anywhere, since you returned the old `GERReader` logic for `aggsender`.

Reply:
It is easier to expose this function and [test](https://github.com/agglayer/aggkit/blob/3c8dc20afaca8fc83343126febb49708b6bb6279/l2gersync/e2e_test.go#L160) the legacy evm downloader that way, that's why I have opted to keep it (and who knows we might need it in future as well).